modules/nw-http2-haproxy.adoc

```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+// * networking/ingress-operator.adoc
+
+[id="nw-http2-haproxy{context}"]
+= HTTP/2 Ingress connectivity
+
+You can enable transparent end-to-end HTTP/2 connectivity in HAProxy. It allows application owners to make use of HTTP/2 protocol capabilities, including single connection, header compression, binary streams, and more.
+
+You can enable this HTTP/2 connectivity for an individual Ingress Controller or for the entire cluster.
+
+To enable the use of HTTP/2 for the connection from the client to HAProxy, a
+route must specify a custom certificate. A route that uses the default
+certificate cannot use HTTP/2. This restriction is necessary to avoid problems
+from connection coalescing, where the client re-uses a connection for different
+routes that use the same certificate.
+
+The connection from HAProxy to the application Pod can use HTTP/2 only for
+re-encrypt routes and not for edge-terminated or insecure routes. This
+restriction comes from the fact that HAProxy uses Application-Level Protocol
+Negotiation (ALPN), which is a TLS extension, to negotiate the use of HTTP/2
```

Comment:
@Miciah, if I'm honest, I have to admit I don't why ALPN means only passthrough and re-encrypt are permissible, and insecure and edge-terminated are not. But I haven't configured an HTTP proxy since v1.1 days, so it might be obvious.

Reply:
Some of the limitation comes from plumbing the correct "server" line in haproxy. 

In https://github.com/openshift/router/pull/104 we were trying to make gRPC transparently work. If you have a client(gRPC)->haproxy(edge terminated)->backend you need to explicitly change the haproxy behaviour away from ALPN. What would normally be `alpn h2,http/1.1` now must be `proto h2`. The following link [1] details the `proto h2` reasoning. 

We really don't have complete information on what the backend does (or can do) to decide whether the config can default to `alpn h2` on edge-terminated routes. Substituting for `proto h2` by automatically sniffing for `application/grpc` in a header works for the specific gRPC case, but breaks down if you are multiplexing on the backend for, say, gRPC **and** HTTP protocols.

The TL;DR is we didn't find a generic 100% accurate mapping for the haproxy config. The current/future thinking is for the service/route to explicitly state its needs and plumb the haproxy config based on that. Overall, the limitation of only supporting re-encrypt and passthrough may be a limitation of configuring haproxy within the context of openshift; I haven't looked at other proxy implementations to see if this could be easier, or more transparent there.

[1] - https://www.haproxy.com/blog/haproxy-2-0-and-beyond/#end-to-end-http-2

Reply:
The reason why the use of ALPN means only it only works for passthrough and reencrypt routes is in the parenthetical ("which is a TLS extension"), but it certainly could be stated more explicitly.

Generally, HTTP/2 is designed for use with TLS, and HAProxy uses ALPN during the TLS handshake to negotiate the use of HTTP/2.  When one uses a passthrough route, HAProxy passes the client connection through to the application, the client and application perform the TLS handshake, and so they can negotiate HTTP/2 using ALPN.  When one uses a reencrypt route, HAProxy opens a TLS connection to the application, HAProxy and the application perform the TLS handshake, and so they likewise can negotiate HTTP/2 using ALPN.  In the case of insecure and edge-terminated routes however, HAProxy opens a cleartext HTTP connection to the application, so there is no TLS and therefore no TLS handshake, no ALPN, and no way to negotiate HTTP/2.